Adds proc support for :allow_blank and :allow_nil

gucki opened the pull request:
Allows to pass a proc to :allow_blank and :allow_nil for validators.

gucki committed:
Adds proc support for :allow_blank and :allow_nil

rafaelfranca commented:
Could you explain what is your use case?

gucki commented:
Here's a simple use case:

- Allow email to be empty if phone number given
- If email is not blank: make sure it is an email, make sure it has a max length of 40 chars

    validates :email, presence: true, length: 1..40, email: true, allow_blank: lambda{ |m| m.phone_number.present? }

rafaelfranca commented on activemodel/test/cases/validations/with_validation_test.rb:
```diff
@@ -142,14 +142,46 @@ def check_validity!
   end
 
+  test "each validator skip nil values if :allow_nil is a proc and returns true" do
+    Topic.validates_with(ValidatorPerEachAttribute, attributes: [:title, :content], allow_nil: lambda{ |record| true })
```
Styles on this file is wrong. Could you run `rubocop -a` on this file?

rafaelfranca reviewed (changes requested):
Ok. It makes sense. Could you add a CHANGELOG entry and also update the documentation?

sgrif commented:
Why is this needed when you can already pass a proc to the `:if` option?

gucki commented:
@rafaelfranca Thanks, I'll post the changes shortly.

@sgrif Imo it's easier to read/ write/ understand using allow_blank than using if:

    validates :email, presence: true, length: 1..40, email: true, if: lambda{ |m| m.email.present? || m.phone_number.blank? }

kaspth commented:
The `m.email.present?` seems redundant to me, so already there's some character savings there.

Here's a different take using Rails' autogenerated attribute predicate methods and the existing validation API:

```ruby
validates :email, presence: true, length: 1..40, email: true, unless: :phone_number?
```

I think that should work for you. Let's revisit this if it doesn't 😄

rafaelfranca commented:
Pay attention that this feature was only disabling the blank check in the validation but the validation was still being applied. With if or unless the validation is not applied at all.

kaspth commented:
Ah, good point. But why not put that on two lines then? I find it really tough to grasp that the `allow_blank` was really only meant for the presence validation, yet grouped in with all the other validations.

Is this what the validation is meant to work?

```ruby
validates_presence_of :email, unless: :phone_number?
validates :email, length: 1..40, email: true
```

gucki commented:
I'm sorry, but your last example would not work as intended. Here are three possible options:

    1.
    validates :email, presence: true if: lambda{ |m| m.phone_number.blank? }
    validates :email, length: 1..40, email: true, if: lambda{ |m| m.email.present? }

    2.
    validates :email, presence: true, length: 1..40, email: true, if: lambda{ |m| m.email.present? || m.phone_number.blank? }

    3.
    validates :email, presence: true, length: 1..40, email: true, allow_blank: lambda{ |m| m.phone_number.present? }

Don't you think option 3 is the easiest to read/ write/ understand?

gucki committed:
Changes to coding style of rubocop for the new tests

gucki committed:
Updates documentation

gucki committed:
Updates changelog

gucki commented:
I just made the requested additions/ changes and rebased to current master. 

I didn't run rubocop on the whole file because that results in "1 file inspected, 90 offenses detected, 74 offenses corrected". I just updated my new tests - let me know if I should commit the whole file.

gucki committed:
Fixes changelog

kaspth commented:
> Don't you think option 3 is the easiest to read/ write/ understand?

Spent 2 comments trying to get it, so not really 😄 — but @rafaelfranca is in favor of this so I'll bow out 👍

carlosantoniodasilva commented:
Thanks for your contribution.

Just adding a couple notes from my understanding: when reading this bit of description/use case:

```
- Allow email to be empty if phone number given
- If email is not blank: make sure it is an email, make sure it has a length between 1 and 40 chars
```

I automatically think of it as two separate validation steps:

```ruby
validates :email, presence: true, unless: :phone_number?
validates :email, length: 1..40, email: true, if: :email?
```

The two validations describe exactly the use case you provide, and seem very easy to follow. It feels a little too complex or maybe more convoluted to bundle everything into a single validation:

```ruby
validates :email, presence: true, length: 1..40, email: true, allow_blank: -> m { m.phone_number? }
```

Not saying I'm totally against it, just thinking for someone reading the requirements and the code, having separate validations seem a little easier to grasp for the eyes.

rafaelfranca commented:
This is a good point. Since we have three contributors with the opinion that this may be harder to understand than separated validations I'm closing this PR.

Thank you for the contribution.

rafaelfranca commented:
Pay attention that this feature was only disabling the blank check in the
validation but the validation was still being applied. With if or unless
the validation is not applied at all.

On Sun, Oct 23, 2016, 08:42 Kasper Timm Hansen <notifications@github.com>
wrote:

> Closed #26811 <https://github.com/rails/rails/pull/26811>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/26811#event-833037296>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAC66J1-o41KGZwf0DpuCuCFonDrTVEOks5q20gYgaJpZM4KZhiD>
> .
>

kaspth commented:
> Pay attention that this feature was only disabling the blank check... but the validation was still being applied

Isn't allowing a blank value in a presence validation the same thing as disabling it? I just don't get this 😋

gucki commented:
@kaspth You are thinking of it way to complex, it just as simple as:
1. If attribute is present all validators are run.
2. If attribute is blank all validators are skipped.

@carlosantoniodasilva 
You are right, it was a bad example *g*. Now what if I fomulate the require as "Make sure email present, is an email and it has a length between 1 and 40 chars, but allow it to be blank if a phone number is given.". It's exactly the same and now reads 1:1 with my validation.

@rafaelfranca I'd love to see this really minor enhancement (3 lines of code changed) in rails. If you don't want to use it, use it otherwise leave it. But if you want to use it and can't, it feels bad and not inline with the other options like `if' and `unless` where you can pass a proc.

kaspth commented:
Sorry, I think we're moving on to look into other things. Though you might not have convinced enough of us with this change, feel free to propose any other in the future. Take care and have a great day! ❤️
